Add support for Dropbox response_type and token_access_type params

kiddailey opened the pull request:
<!--

Summarize the changes this Pull Request makes.

Please include a reference to a GitHub issue if appropriate.
It is not necessary to create a new issue just to open a Pull Request.

If adding a new provider, please ensure you have included appropriate test(s).

For new providers, a package with the same ID must not already exist in NuGet.org.
If it does, you must be the owner of the package and agree to transfer ownership
of the package to the [aspnet-contrib](https://www.nuget.org/profiles/aspnet-contrib)
organisation after the Pull Request is merged so we can take over publishing of
the provider from our Continuous Integration.

-->

This pull request address Feature Request #578 and adds support for the Dropbox token_access_type and response_type parameters:

**ResponseType** allows specification of the grant type. Per the [documentation](https://www.dropbox.com/developers/documentation/http/documentation):

> The grant type requested, either token or code. 

**AccessType** allows for the ability to request offline access and receive refresh_tokens with authentication tokens.   Per the [documentation](https://www.dropbox.com/developers/documentation/http/documentation):

>  If this parameter is set to offline, then the access token payload returned by a successful /oauth2/token call will contain a short-lived access_token and a long-lived refresh_token that can be used to request a new short-lived access token as long as a user's approval remains valid. If set to online then only a short-lived access_token will be returned. If omitted, the response will default to returning a long-lived access_token if they are allowed in the app console. If long-lived access tokens are disabled in the app console, this parameter defaults to online

kiddailey committed:
Add support for Dropbox response_type and token_access_type params

martincostello commented on src/AspNet.Security.OAuth.Dropbox/DropboxAuthenticationOptions.cs:
```diff
@@ -15,6 +15,16 @@ namespace AspNet.Security.OAuth.Dropbox
     public class DropboxAuthenticationOptions : OAuthOptions
     {
+        /// <summary>
+        /// Gets or sets what the response type from Dropbox should be:  Online, Offline or Legacy
```
```suggestion
        /// Gets or sets what the response type from Dropbox should be: Online, Offline or Legacy.
```

martincostello commented on src/AspNet.Security.OAuth.Dropbox/DropboxAuthenticationOptions.cs:
```diff
@@ -15,6 +15,16 @@ namespace AspNet.Security.OAuth.Dropbox
+        public string? AccessType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Response Type requested from Dropbox:  Token or Code
```
```suggestion
        /// Gets or sets the response type requested from Dropbox: Token or Code.
```

martincostello reviewed (commented):
Thanks for the contribution 👍

Just a few extra things needed:

1. Please add a `dropbox.md` file to the `docs` folder for the additional parameters based on the format used by the existing provider documentation.
2. Please add at least one additional test for the Dropbox provider using both of the new options to cover the new code paths.

Could you also confirm you've tested this works as expected with the real Dropbox APIs?

kevinchalet commented:
@kiddailey thanks for your PR!

It's worth noting that `response_type=token` cannot work with server-side clients unless the OAuth 2.0 server supports the `response_mode=form_post` extension, as `response_type=token` implies returning the token material as part of the URL fragment, which is not sent by browsers to HTTP servers. I don't think Dropbox supports `response_mode=form_post` and if even if it did, it's not something the OAuth 2.0 base handler supports anyway (and it doesn't even support the implicit flow).

Personally, I'd remove the `ResponseType` thing and focus on the `AccessType` addition. We could even go one step further and set it to `offline` to encourage developers to use refresh tokens instead of the long-lived access tokens that were deprecated by Dropbox (which is a good thing).

kiddailey commented:
Thanks for the great feedback!

@martincostello: I'll take care of adding these as soon as I can.  I have tested it partially against Dropbox.  Specifically, in a Xamarin app using XamarinEssential's WebAuthenticator with my own server facilitating.  Everything appears to work as expected, but I still need to finish testing and implement the refresh_token handling, which I should get to today.  In any case, Dropbox appears to be responding as expected.

Edit: Confirmed my previous tests that I can receive tokens and refresh tokens as expected with code/offline and can authenticate with them.   ~~Testing the fetch of a new token using the provided refresh is still pending.~~ The Dropbox API reportedly automatically handles renewing the token if refresh token is provided on instantiation.  ~~Testing to confirm with my use case (takes 4 hours for short-lived tokens to expire).~~ Yes, the refresh token I received was able to be used to authenticate again.

@kevinchalet: Agreed.  As I specifically needed code flow and offline refresh tokens, I mainly added this so that I didn't break existing behavior.  Would it be acceptable to include the information you mention, note the deprecation in the readme, and leave this in, or would it be preferred to note the deprecation and make offline a constant?

kiddailey committed:
Follow comment standards in src/AspNet.Security.OAuth.Dropbox/DropboxAuthenticationOptions.cs

Co-authored-by: Martin Costello <martin@martincostello.com>

kiddailey committed:
Add Dropbox readme with example and optional parameter documentation

kiddailey committed:
Fix casing on options comment

kiddailey committed:
Remove ResponseType option

kiddailey committed:
Add token_access_type tests

kiddailey commented:
Readme and tests added.  And to confirm from my earlier post, I have tested this against Dropbox actual in an "offline" app that refreshes tokens as needed and it works as expected.

You'll probably note that I removed the ResponseType support per kevinchalet's suggestion.  In writing the tests and looking more at things, I realized that the response_type was already being set to "code" (obviously) and that "token" was likely not going to work properly without more effort.   For those reasons, I went ahead and removed it as suggested.

If the consensus is to make "offline" the default, let me know and I'll make that change as well -- and probably note it in the readme as it may break things for some.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/Dropbox/DropboxTests.cs:
```diff
@@ -42,5 +45,71 @@ public async Task Can_Sign_In_Using_Dropbox(string claimType, string claimValue)
+                    {
+                        OnRedirectToAuthorizationEndpoint = ctx =>
+                        {
+                            accessTypeIsSet = ctx.RedirectUri.Contains($"token_access_type={value}", StringComparison.InvariantCulture);
```
```suggestion
                            accessTypeIsSet = ctx.RedirectUri.Contains($"token_access_type={value}", StringComparison.OrdinalIgnoreCase);
```

martincostello reviewed (commented):
Just one minor change. Otherwise looks good to me.

Let's not change the default behaviour as I'd consider that to be a breaking change that we wouldn't do outside of a major release.

kiddailey committed:
Ignore case in string comparison of test

Co-authored-by: Martin Costello <martin@martincostello.com>

kevinchalet reviewed (approved):
Looks great, thanks!

Do you think it's worth introducing a `DropboxAuthenticationConstants` class containing a nested `TokenAccessTypes` class with online/offline/legacy constants?

kiddailey commented:
> Do you think it's worth introducing a `DropboxAuthenticationConstants` class containing a nested `TokenAccessTypes` class with online/offline/legacy constants?

That is a great question!  I originally did write it that way, but when looking at the style of the other providers, it seemed that many were using simple strings so I went with that.

Personally I do prefer doing it with constants, though that does remove a little bit of flexibility should the dropbox API change before the provider can be updated.  Let me know if you'd like me to change it and I'll take care of it.

Thanks!

martincostello commented:
I think for now we can just take this as-is, rather than add some extra constants, unless anyone feels strongly the other way.

martincostello commented:
This change is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.Dropbox/5.0.9
